### DISCUSSION TITLE
Shorter way to watch future provider

### DISCUSSION THREAD

**Discussion by Colman** [rrousselGit/riverpod#3434] [Q&A]
In many places in my app, I have a future provider that depends on another future provider. For example:

```dart
@riverpod
Future<RegionModel> region(RegionRef ref) async {
  final someParam = await ref.watch(someParamProvider.future);
  
  final result = await ref.watch(someResultProvider(someParam).future);
  //More calculations
}
```

With region model defined as:

```dart
@immutable
class RegionModel {
  const RegionModel({
    required this.currentBranchId,
    required this.currentCityId,
    required this.lastDefinedBranchId,
    required this.lastDefinedCityId,
  });

  final int? currentBranchId;
  final int? currentCityId;
  final int lastDefinedBranchId;
  final int lastDefinedCityId;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RegionModel &&
          runtimeType == other.runtimeType &&
          _currentBranchId == other._currentBranchId &&
          _currentCityId == other._currentCityId &&
          _lastDefinedBranchId == other._lastDefinedBranchId &&
          _lastDefinedCityId == other._lastDefinedCityId;

  @override
  int get hashCode =>
      _currentBranchId.hashCode ^
      _currentCityId.hashCode ^
      _lastDefinedBranchId.hashCode ^
      _lastDefinedCityId.hashCode;
}
```

Every time `someParamProvider` reloads, the region provider also runs again, even if the value of someParam didn't change after the reload. When you watch the future, shouldn't it only trigger an update when the actual value changes?

As a workaround to this issue, you can select the value like so:
```dart
@riverpod
Future<RegionModel> region(RegionRef ref) async {
  final someParam = await ref.watch(someParamProvider.selectAsync((value) => value);
  
  final result = await ref.watch(someResultProvider(someParam).future);
  //More calculations
}
```

This will only rebuild region when the actual value of `someParam` changes. I think this is what developers want in most cases. So shouldn't it be the default behavior of watching the future? It's quite tedious to call `selectAsync` everywhere when you are selecting the entire model. If this cannot be made to the default behavior, could a shortcut be added like this `someParamProvider.selectValue`? This would be equivalent to `someParamProvider.selectAsync((value) => value)`.

**Comment by rrousselGit** [comment]
>  I think this is what developers want in most cases. So shouldn't it be the default behavior of watching the future? 

Not necessarily.
The problem with selectAsync is, you're essentially delaying your second request until the first one has completed. This is waterfall-like.  
The current behaviour enables starting your secondary request earlier. 


> It's quite tedious to call selectAsync everywhere when you are selecting the entire model. 

Usually you shouldn't be selecting "the entire model". Pick only what you need in that model. Maybe in your specific scenario you need to, but that's not supposed to be common.

And you can make an extension method to help if you want:

```dart
extension<T> on FutureProvider<T> {
   ProviderListenable<T> get myFuture => selectAsync((a) => a); 
}
```

**Reply by Colman** [reply]
> The problem with selectAsync is, you're essentially delaying your second request until the first one has completed. This is waterfall-like. The current behavior enables starting your secondary request earlier.

But the second request can't start anyway until it has the value from the first one since its value depends on the param. That's why there's an `await` on the `someParamProvider`. I've edited my example to make it a bit more clear that there's a dependency on the param provider.

> Usually you shouldn't be selecting "the entire model". Pick only what you need in that model.

In my example, I didn't intend for `someParam` to be a model. It would be an int or some other primitive type. So I should've said "selecting the value" instead of "the entire model". Sorry about that. Even if it was a model, selecting the whole model is quite common if the fields in the model usually change at the same time. Even the docs say that select may not be worth it if the properties rarely change individually. https://riverpod.dev/docs/advanced/select#:~:text=Using%20select%20slightly%20slows%20down%20invididual%20read%20operations%20and%20increase%20a%20tiny%20bit%20the%20complexity%20of%20your%20code.%0AIt%20may%20not%20be%20worth%20using%20it%20if%20those%20%22other%20properties%22%20rarely%20change.

What would be the case where you want to watch changes to the future itself, instead of just the value that the future gives? And would that case be more common than my example? If you really did want `regionProvider` to update when the param future itself updates, couldn't you achieve that by doing this:

```dart
ref.watch(someParamProvider).isLoading;
final someParam = await ref.watch(someParamProvider.selectAsync((value) => value);

final result = await ref.watch(someResultProvider(someParam).future);
```

**Reply by rrousselGit** [reply]
> But the second request can't start anyway until it has the value from the first one since its value depends on the param. That's why there's an await on the someParamProvider. I've edited my example to make it a bit more clear that there's a dependency on the param provider.

_Your_ request can't. That doesn't mean all requests can't.
A FutureProvider may do more asynchronous stuff than just a single `await ref.watch(someParamProvider.future)`.

There's also the fact that we want to notify the UI as soon as possible that the provider is busy.
If you use `selectAsync` and your UI only watches the second request, you're essentially delaying the spinner in the UI until the first request has completed.

> Even the docs say that select may not be worth it if the properties rarely change individually.

That note is with regards to updating Widgets/UI. Async providers are a different beast.
For FutureProviders, I think it's fine to use selectors often.


---

Also note that codegen will have some sugar with regards to selectors.
I'll offer things like:


```dart
class Model {
  String name;
}

@riverpod
Future<Model> model(...) {...}

AsyncValue<String> a=  ref.watch(modelProvider.name); // sugar for select((v) => v.name);
Future<String> b = ref.watch(modelProvider.name.future); // sugar for selectAsync((v) => v.name);
```

**Reply by Colman** [reply]
> Your request can't. That doesn't mean all requests can't.

I understand that it doesn't mean that all requests can't. I wasn't saying that. I just think that it's a lot _more common_ that when a future provider watches another future provider, it cares about its value not the future. Do you agree with this? Based on the assumption that it's the more common case, I would think there's a shorter/easier way to do it.

Regarding your code snippet, you say that:
```dart
Future<String> b = ref.watch(modelProvider.name.future); // sugar for selectAsync((v) => v.name);
```

So if you follow that pattern, wouldn't it be logical that if you omit "name" from the code, it should select the whole model instead of just the name param?
```dart
Future<String> b = ref.watch(modelProvider.future); // sugar for selectAsync((v) => v);
```

Why is it that when you watch `.future` on "name" it reloads the watching provider AFTER the modelProvider loads? But then when you use `.future` without "name", it reloads the watching provider when `modelProvider` STARTS loading and AFTER it loads. It just seems inconsistent to me.


> There's also the fact that we want to notify the UI as soon as possible that the provider is busy.
If you use selectAsync and your UI only watches the second request, you're essentially delaying the spinner in the UI until the first request has completed.

This is a good point and I've done some thinking about it. For the initial load of the second provider, I think the spinner will show as it's supposed to since the second provider has no value until the `selectAsync` inside of it resolves. As for the reloads after that, shouldn't this be something that Riverpod handles? If you use:

```dart
@riverpod
Future<ModelB> modelB(ModelBRef ref) async {
   final name = await ref.watch(modelAProvider.selectAsync((modelA) => modelA.name));
   //...
}
```

Could Riverpod update the `isReloading` flag for `modelBProvider` when `modelAProvider` starts loading instead of when `modelA.name` changes? Without actually rerunning the body of provider B since name hasn't changed yet.

I know my idea is a bit convoluted, but currently you have to choose between either optimizing your future providers by using `selectAsync` or having the `isReloading` flag be correct for the UI. It doesn't seem like you can have both.
